File: 0430_flatten_multilevel_linked_list/flatten.go
package main

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}

// Time: O(n)
// Space: O(n)
func flatten(root *Node) *Node {
	if root == nil {
		return nil
	}
	s := Node{0, nil, nil, nil}
	stack := []*Node{root}
	prev := &s
	for len(stack) > 0 {
		curr := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		prev.Next = curr
		curr.Prev = prev
		if curr.Next != nil {
			stack = append(stack, curr.Next)
			curr.Next = nil
		}
		if curr.Child != nil {
			stack = append(stack, curr.Child)
			curr.Child = nil
		}
		prev = curr
	}
	s.Next.Prev = nil
	return s.Next
}

func main() {

}

File: 0430_flatten_multilevel_linked_list/flatten.py
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child


def flatten(head: 'Node') -> 'Node':
    """
    Time: O(n)
    Space: O(n)
    """
    if not head:
        return
    s = Node(0, None, None, None)
    prev = s
    stack = [head]
    while stack:
        curr = stack.pop()
        curr.prev = prev
        prev.next = curr
        if curr.next:
            stack.append(curr.next)
            curr.next = None
        if curr.child:
            stack.append(curr.child)
            curr.child = None
        prev = curr
    s.next.prev = None
    return s.next
